Accept --template-dir as an argument

kirs opened the pull request:
This is required to specify custom directory with k8s templates. It's `config/deploy/$ENVIRONMENT` by default, but for Shopify core it has to be another path. Right now the deploy fails because it can't find any k8s configs in `config/deploy/$ENVIRONMENT`.

closes https://github.com/Shopify/kubernetes-deploy/pull/6

review @sirupsen @KnVerey

kirs commented on lib/kubernetes-deploy/runner.rb:
```diff
@@ -55,11 +55,12 @@ def self.with_friendly_errors
       @current_sha = current_sha
-      @template_path = File.expand_path('./' + (template_folder || "config/deploy/#{environment}"))
+
+      @template_dir = Pathname.new(template_dir).realpath
```
@KnVerey this way we can accept both absolutely and relative paths; relative path will be converted to absolute.

sirupsen replied:
What is the difference between `Pathname.new(..)` and `File.expand_path`?

```ruby
irb(main):007:0> Pathname.new("~/.bashrc").realpath
Errno::ENOENT: No such file or directory @ realpath_rec - /Users/sirup/~
	from (irb):7:in `realpath'
	from (irb):7:in `realpath'
	from (irb):7
	from /opt/rubies/2.3.3/bin/irb:11:in `<main>'
irb(main):008:0> File.expand_path("~/.bashrc")
=> "/Users/sirup/.bashrc"
```

In this I feel `File.expand_path` wins 🤔  But for anything not involving `~`, it seems to work the same.

KnVerey replied:
`File.expand_path` does look preferable given those examples, but really I'm happy as long as it is possible to write `template_dir: config/my/custom/path` (relative, no tilde) into your shipit.yml.

sirupsen commented on exe/kubernetes-deploy:
```diff
@@ -15,18 +15,23 @@ require 'kubernetes-deploy'
 
+if !template_dir || template_dir.empty?
+  raise ArgumentError, "--template-dir flag is required"
+end
```
Do you think this is required? The default of `config/deploy` makes sense to me 🥇

KnVerey replied:
Agreed, this should not be a required arg and the default should remain on this side.

kirs replied:
The thing is that we want the default to be `config/deploy/$ENVIRONMENT`, so the script would need `environment` argument just to construct the default template dir. IMO adding extra argument (environment) only to construct the template dir is not the best design.

This is the reason why I moved the default `--template-dir` construction to Shipit side where $ENVIRONMENT is available: https://github.com/Shopify/shipit-engine/pull/640/files#diff-01916dce352e57dea198b45614795a9dR17

KnVerey replied:
I see your point, though I'm not convinced that the default belongs in the other repo more than it does here. You could also eliminate the arg to `Runner.new` by setting the default right here, no? I don't feel super strongly about this though.

sirupsen replied:
Ah. Can we remove the need for `ENVIRONMENT` altogether then?

sirupsen commented on exe/kubernetes-deploy:
```diff
@@ -15,18 +15,23 @@ require 'kubernetes-deploy'
+template_dir = nil
 ARGV.options do |opts|
   opts.on("-w", "--skip-wait") { skip_wait = true }
+  opts.on("-tDIR", "--template-dir=DIR") { |v| template_dir = v }
```
I wonder if we should have the shorthands (`-w` and `-t`). We'd always spell them out in scripts, and this shouldn't really be run by hand—even then it's prone to error. Should we consider removing them?

kirs commented:
Thanks for review! I updated the code to use `File.expand_path` and removed shorthands of CLI arguments because I agree that we don't really need them. Please let me know what you think about setting default template dir on the Shipit side.

KnVerey commented:
That variable is always set by shipit in all deploys.

On Jan 30, 2017 19:01, "Simon Eskildsen" <notifications@github.com> wrote:

> *@sirupsen* commented on this pull request.
> ------------------------------
>
> In exe/kubernetes-deploy
> <https://github.com/Shopify/kubernetes-deploy/pull/11>:
>
> >    opts.parse!
>  end
>
> +if !template_dir || template_dir.empty?
> +  raise ArgumentError, "--template-dir flag is required"
> +end
>
> Ah. Can we remove the need for ENVIRONMENT altogether then?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Shopify/kubernetes-deploy/pull/11>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEkU9fU26q5uvJDHrekzJMi48a_EaFvmks5rXnmFgaJpZM4LwjBh>
> .
>

sirupsen commented:
> That variable is always set by shipit in all deploys.

Right! I was thinking that this script might not need to depend on it anymore if `--template-dir` was the last thing that required it. Could it makes sense to do the following:

1. If `--template-dir` is passed, used whatever is passed.
2. If `$ENVIRONMENT` is passed (always true for 🚢 as Katrina pointed out), use the `config/deploy/$ENVIRONMENT` default (and make this a strong convention).
3. Else, error asking people to either provide `$ENVIRONMENT` to infer the template directory or `--template-dir`

👍 👎 ❓

kirs commented:
Great suggestion, Simon! I implemented the idea.
Please 👀 when you have a minute because I'd like to ship this ASAP for the undercloud team.

KnVerey commented on lib/kubernetes-deploy/runner.rb:
```diff
@@ -201,10 +202,10 @@ def validate_configuration
-      elsif Dir.entries(@template_path).none? { |file| file =~ /\.yml(\.erb)?$/ }
-        errors << "#{@template_path} doesn't contain valid templates (postfix .yml or .yml.erb)"
+      if !File.directory?(@template_dir)
+        errors << "Template path #{@template_dir} doesn't exist"
```
"Template directory" in the error message for consistency?

kirs replied:
Thanks, fixed!
